Pause the deployment config before deleting

mfojtik opened the pull request:
Fixes: https://github.com/openshift/origin/issues/9532

kargakis commented on pkg/deploy/reaper/reaper.go:
```diff
@@ -91,3 +114,21 @@ func (reaper *DeploymentConfigReaper) Stop(namespace, name string, timeout time.
+
+type updateDeploymentConfigFunc func(d *deployapi.DeploymentConfig)
+
+func (reaper *DeploymentConfigReaper) updateDeploymentWithRetries(namespace, name string, applyUpdate updateDeploymentFunc) (deployment *deployapi.DeploymentConfig, err error) {
```
Remove the named returns

kargakis commented on pkg/deploy/reaper/reaper.go:
```diff
@@ -91,3 +114,21 @@ func (reaper *DeploymentConfigReaper) Stop(namespace, name string, timeout time.
+		}
+		return false, nil
+	})
+	return deployment, err
```
s/deployment/config/

kargakis commented on pkg/deploy/reaper/reaper.go:
```diff
@@ -30,6 +32,27 @@ type DeploymentConfigReaper struct {
 // replication controller and its deployment configuration.
 func (reaper *DeploymentConfigReaper) Stop(namespace, name string, timeout time.Duration, gracePeriod *kapi.DeleteOptions) error {
+	// Pause the deployment config marked for deletion first.
+	deployment, err := reaper.updateDeploymentWithRetries(namespace, name, func(d *deployapi.DeploymentConfig) {
```
s/deployment/config/

kargakis commented on pkg/deploy/reaper/reaper.go:
```diff
@@ -30,6 +32,27 @@ type DeploymentConfigReaper struct {
+		if err != nil {
+			return false, err
+		}
+		return dc.Status.ObservedGeneration >= deployment.Generation
```
You need to keep the generation from the deployment config you got above and use that instead of using the generation of the object you get every time.

mfojtik commented:
@kargakis thanks! all fixed.

kargakis commented:
[test]

kargakis commented:
Actually can you also update the extended test for the reaper to run in the conformance suite (if it's not already updated) ?

mfojtik commented:
@kargakis will do

kargakis commented on pkg/deploy/reaper/reaper.go:
```diff
@@ -39,8 +66,8 @@ func (reaper *DeploymentConfigReaper) Stop(namespace, name string, timeout time.
 	// for more details.
-	err := reaper.oc.DeploymentConfigs(namespace).Delete(name)
-	configNotFound := kerrors.IsNotFound(err)
+	err = reaper.oc.DeploymentConfigs(namespace).Delete(name)
```
The deletion of the dc needs to move in the end

kargakis commented:
> @kargakis will do

https://github.com/openshift/origin/blob/d26c9ce221cd9cfc986faa4fcd2a8d046394178b/test/extended/deployments/deployments.go#L483

mfojtik commented on pkg/deploy/reaper/reaper.go:
```diff
@@ -26,26 +29,44 @@ type DeploymentConfigReaper struct {
 
+// pause marks the deployment configuration as paused to avoid triggering new
+// deployments.
+func (reaper *DeploymentConfigReaper) pause(namespace, name string) (*deployapi.DeploymentConfig, error) {
```
@kargakis I would love to have `Pause()` func in deployment utils ;-)

kargakis commented on pkg/deploy/reaper/reaper.go:
```diff
@@ -84,10 +106,34 @@ func (reaper *DeploymentConfigReaper) Stop(namespace, name string, timeout time.
+		// Apply the update, then attempt to push it to the apiserver.
+		applyUpdate(config)
+		config, err = deploymentConfigs.Update(config)
+		if err != nil {
```
We should retry only on update conflicts?

mfojtik replied:
@kargakis this is from upstream, if we want that I will fix this in upstream as well.

kargakis replied:
Yes, we should probably fix it there too

mfojtik replied:
fixed (here)

mfojtik replied:
And the upstream portion: https://github.com/kubernetes/kubernetes/pull/29188

kargakis commented:
LGTM [merge]

mfojtik commented on pkg/deploy/reaper/reaper.go:
```diff
@@ -26,26 +29,58 @@ type DeploymentConfigReaper struct {
+		// If we failed to pause the deployment config, it means we are talking to
+		// old API that does not support pausing. In that case, we delete the
+		// deployment config to stay backward compatible.
+		if !isPaused {
```
@kargakis backward compatibility should be saved

kargakis replied:
sup

kargakis commented on pkg/deploy/reaper/reaper.go:
```diff
@@ -62,10 +97,11 @@ func (reaper *DeploymentConfigReaper) Stop(namespace, name string, timeout time.
 		if err = rcReaper.Stop(rc.Namespace, rc.Name, timeout, gracePeriod); err != nil {
 			// Better not error out here...
-			glog.Infof("Cannot delete ReplicationController %s/%s: %v", rc.Namespace, rc.Name, err)
+			glog.Infof("Cannot delete ReplicationController %s/%s for deployment %s/%s: %v", rc.Namespace, rc.Name, namespace, name, err)
```
for deployment config

mfojtik commented:
@kargakis this now need re-review

1. Added legacy behavior, so when we fail to set the `Pause=true`, we delete the deployment config at the top
2. Added tests for legacy behavior, so we know both pausing and legacy works together

mfojtik commented:
[test]

kargakis commented on pkg/deploy/cmd/reaper.go:
```diff
@@ -83,10 +121,44 @@ func (reaper *DeploymentConfigReaper) Stop(namespace, name string, timeout time.
+		err    error
+	)
+	deploymentConfigs := reaper.oc.DeploymentConfigs(namespace)
+	resultErr := wait.Poll(10*time.Millisecond, 1*time.Minute, func() (bool, error) {
```
Don't poll too fast - make this 500ms.

mfojtik replied:
this was copied from upstream :)

kargakis commented:
One comment, LGTM otherwise

kargakis commented:
LGTM [merge]

mfojtik commented:
[merge]

mfojtik commented:
[merge] (RH registry flake)
